openstack_db_database_v1 resource

eugenetaranov opened the pull request:
This PR adds openstack_db_database_v1 resource and addresses https://github.com/terraform-providers/terraform-provider-openstack/issues/141

This is a contribution from [Ribose Inc.](https://www.ribose.com)

jtopjian commented:
oops - sorry, commented on the wrong PR earlier :)

@eugenetaranov This looks good, though I noticed an oddity. It might just be my environment, though.

Trove is only returning the database name when querying the databases of an instance:

2017/12/06 04:26:18 [DEBUG] OpenStack Response Body: {
  "databases": [                                      
    {                                                 
      "name": "basic"                                 
    }                                                 
  ]                                                   
}                                                     

Because of this, collate and charset are being set to empty strings which is causing a conflict with the values given in the test fixture.

Can you confirm if these values are returning for you?

You can view extra log information by setting the OS_DEBUG environment variable to 1.

eugenetaranov commented:
@jtopjian checked, looks like trove does not send anything besides the db names indeed:

`
2017/12/06 14:43:28 [DEBUG] Openstack Response Code: 200
2017/12/06 14:43:28 [DEBUG] Openstack Response Headers:
Content-Length: 51
Content-Type: application/json
Date: Wed, 06 Dec 2017 11:43:28 GMT
Server: nginx/1.10.3 (Ubuntu)
2017/12/06 14:43:28 [DEBUG] OpenStack Response Body: {
  "databases": [
    {
      "name": "basic"
    },
    {
      "name": "sys"
    }
  ]
}
2017/12/06 14:43:28 Retrieved databases [{basic  } {sys  }]
`
Confirmed also with mitmproxy.

jtopjian commented:
@eugenetaranov OK, maybe that was a change in recent versions of Trove? Or it's an error in Gophercloud.

For now, I recommend not setting collate and charset via the API. During the Read call, just confirm the database name was returned.

eugenetaranov commented:
@jtopjian thank you, removed charset and collate for now

jtopjian commented on openstack/resource_openstack_db_database_v1.go:
```diff
@@ -0,0 +1,176 @@
+	}
+
+	// Store the ID now
+	d.SetId(instance_id)
```
The ID of the resource should be `instance_id/db_name`. If it's just `instance_id`, then this will limit a user from only be able to create one `openstack_db_database_v1` resource per database instance.

jtopjian commented on openstack/resource_openstack_db_database_v1.go:
```diff
@@ -0,0 +1,176 @@
+		return fmt.Errorf("Error creating cloud database client: %s", err)
+	}
+
+	dbname := d.Get("name").(string)
```
If the ID is now `instance_id/db_name`, you can now do:

```go
pieces := strings.Split(db.Id(), "/", 2)
instanceID := pieces[0]
dbname := pieces[1]
```

jtopjian commented on openstack/resource_openstack_db_database_v1.go:
```diff
@@ -0,0 +1,176 @@
+			d.Set("name", v.Name)
+			break
+		}
+	}
```
There should be some type of error check here if the dbname could not be found. Similar to what you're doing in `Delete` with `dbExists`. You should return an error if the database could not be found.

jtopjian commented on openstack/resource_openstack_db_database_v1.go:
```diff
@@ -0,0 +1,176 @@
+	}
+
+	if !dbExists {
+		log.Printf("Database %s was not found on instance %s", dbname, d.Id())
```
In addition to logging, also just `return nil` here. The database doesn't exist and we shouldn't try to do a `databases.Delete` call here.

jtopjian commented on openstack/resource_openstack_db_database_v1_test.go:
```diff
@@ -0,0 +1,90 @@
+				return nil
+			}
+		}
+		return nil
```
I think this should return an error that the database could not be found.

eugenetaranov commented:
@jtopjian thank you, addressed all points, the only thing that I did differently is database ID: chose dot "." as a separator since "/" might cause issues when sent in http request.
In addition, found a bug: terraform.tfstate does not include openstack_db_database_v1 resourcem it is not generated. My test config is below, could you please advise what may cause such behavior.

```resource "openstack_db_instance_v1" "basic" {
  region        = "region"
  name          = "test"
  flavor_id     = "31792d21-c355-4587-9290-56c1ed0ca386"
  size          = 10
  network {
    uuid = "8ece1c9a-0ae9-4ffa-819e-cc5523bacc53"
  }

  datastore {
    version = "mysql-5.7"
    type    = "mysql"
  }
}

resource "openstack_db_database_v1" "test" {
  name     = "testdb18"
  instance = "${openstack_db_instance_v1.basic.id}"
}
```

animationzl commented:
recheck

jtopjian commented:
This looks good to me. @eugenetaranov Thank you for your work on this.

@animationzl let me know when I can merge this. Feel free to use this to test out the CI environment for however long you need.

jtopjian commented:
recheck trove

jtopjian commented:
LGTM - thank you @eugenetaranov  :smile:
